Very WIP: sequential colors for plot_compare_evoked

jona-sassenhagen opened the pull request:
PR early, PR often.

![unknown-240](https://cloud.githubusercontent.com/assets/4321826/25329480/2a05dc9c-28dc-11e7-88a8-9d297096970c.png)

See
http://nbviewer.jupyter.org/gist/jona-sassenhagen/b1fca57f72e36b8c225d0315b18a0206

mmagnuski commented on mne/viz/evoked.py:
```diff
@@ -1723,7 +1755,42 @@ def plot_compare_evokeds(evokeds, picks=list(), gfp=False, colors=None,
+
+        norm = mpl.colors.BoundaryNorm(numbers + [numbers[-1] +1], l.N)
+        sm = matplotlib.cm.ScalarMappable(cmap=l, norm=norm
+                                         )
```
a lonely bracket, lost in the code, help it find its way home!

jona-sassenhagen replied:
That whole chunk of code is just very terrible in general ... I don't dare fixing it yet.

mmagnuski replied:
But you've given some hope to the lonely bracket now ;)

jona-sassenhagen commented:
@mmagnuski do you think you can test it a bit?

jona-sassenhagen commented:
I don't get where the error is coming from:

```
======================================================================
FAIL: Test module docstring formatting.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/miniconda/envs/test/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/travis/build/mne-tools/mne-python/mne/utils.py", line 824, in dec
    return function(*args, **kwargs)
  File "/home/travis/build/mne-tools/mne-python/mne/tests/test_docstring_parameters.py", line 143, in test_docstring_parameters
    raise AssertionError(msg)
AssertionError: 
mne.utils.run_subprocess arg mismatch: ['-------', 'Returns', 'stderr', 'stdout']
>>  raise AssertionError("\nmne.utils.run_subprocess arg mismatch: ['-------', 'Returns', 'stderr', 'stdout']")
    
```

Eric89GXL commented:
@jona-sassenhagen I'll see if I can fix it

Eric89GXL commented on mne/utils.py:
```diff
@@ -1172,8 +1172,8 @@ def run_subprocess(command, verbose=None, *args, **kwargs):
 
-    Returns
     -------
+    Returns
```
@jona-sassenhagen this is where it comes from

jona-sassenhagen replied:
How the hell did that show up here!?

Thanks Eric. I was combing through the parts I could *remember* changing, but couldn't find anything.

Eric89GXL replied:
FWYW on a somewhat related note I find that checking my own diffs top-to-bottom on GitHub I usually find one or two things I didn't mean to put in there or forgot to change. I usually try to do it before requesting a review, as it cuts down on the review iteration cycles/time a bit.

jona-sassenhagen replied:
Ha, this isn't ready for review yet, first I have to see if I can convince @agramfort to allow all of the clumsy magic for the sake of a convenience plotting function.

jona-sassenhagen commented:
Sadly this is adding like 2 secs to the tests.

jona-sassenhagen commented:
Error is just coverage ...

@agramfort did you have time to look at the API (see gist)?
Maybe @dengemann has comments?

mmagnuski commented:
@jona-sassenhagen Yes, I can test later today. I'll let you know.

jona-sassenhagen commented:
Thanks @mmagnuski ... I don't mean 'stress testing', just if you find the function useful and the API sensible.

agramfort commented:
When I set sequential to False in the last cell it breaks.

To me it's too much magic. I would go with a more explicit approach, maybe inspired by pandas or seaborn.

```
meta = dict(num_letters=[4, 4, ...], concrete=[0, 1, 0, 1]) 
# or
meta = pd.DataFrame(dict(num_letters=[4, 4, ...], concrete=[0, 1, 0, 1]))

mne.viz.evoked.plot_compare_evokeds(
    evokeds.values(), meta=meta, color='num_letters', linestyle='concrete',
    colormap=plt.cm.viridis
    picks=...
).set_size_inches((6, 3));
```

inferring variables from strings (keys here) is for me very much like black magic.

jona-sassenhagen commented:
So a design matrix-like approach?

How about:

```
evokeds = {cond1 + "/" + cond2: epochs[cond1 + cond2].average()
    for cond2 in {"freq_1", "_freq_2", "freq_3} for cond1 in {"concrete", "abstract"}}
order = {cond: int(cond.split("_")[-1]) for cond in evokeds}

mne.viz.plot_compare_evokeds(evokeds, colors="viridis", order=order)
```

(E.g. a dict mapping conditions to integers, where the integers indicate color on a colormap)

The API needs the split legend though. Though I think this can be as easy as directly mapping the linestyles dict to a legend.
Zero magic in either case.

One also needs to set the colorbar label though, somehow. E.g., minimally, here

`mne.viz.plot_compare_evokeds(evokeds, colors="viridis", color_units='frequency', order=order)`

agramfort commented:
let's draft the different use cases:

First a 2/2 design:
```
evokeds = {'A/C': evoked1, 'B/C': evoked2, 'A/C': evoked3, 'B/D': evoked4}
mne.viz.plot_compare_evokeds(evokeds, colors="viridis")
```

then a mix of categorical and continuous:
```
evokeds = {'A/1': evoked1, 'A/1': evoked2, 'B/2': evoked3, 'B/2': evoked4}
mne.viz.plot_compare_evokeds(evokeds, colors="viridis")
```
or using meta
```
meta = dict(n_letters=[1, 1, 2, 2])
evokeds = {'A/1': evoked1, 'A/1': evoked2, 'B/2': evoked3, 'B/2': evoked4}
mne.viz.plot_compare_evokeds(evokeds, colors="viridis", meta=meta, color='n_letters')
```

@jona-sassenhagen please edit this so we have a compact view of what use case we want to support

mmagnuski commented:
@jona-sassenhagen sorry for the delay, it works well for me but I tested only simple cases with one sequential var. I like the `meta` approach - but it would require using OrderedDict or list for evokeds.

jona-sassenhagen commented:
@agramfort something changed the post as I was editing and I lost all my comments, so I'm putting it here ...


Factorial:

```
evokeds = {'A/C': evoked1, 'B/C': evoked2, 'A/C': evoked3, 'B/D': evoked4}
mne.viz.plot_compare_evokeds(evokeds, colors="viridis")
or
mne.viz.plot_compare_evokeds(evokeds, colors={"A":"r", "B":"b"}, linestyles={"C":"-", "D":":"})
```

Continuous

```
evokeds = [evoked1, evoked2, evoked3, evoked4]
colors = list(range(len(evokeds)))
cmap = {"colorbar_label":"viridis"}
mne.viz.plot_compare_evokeds(evokeds, colors=colors, cmap=cmap)
```
`colors` is of the same shape as `evokeds`

Mixed:
```
evokeds = {'A/1': evoked1, 'A/2': evoked2, 'B/1': evoked3, 'B/2': evoked4}
linestyles = {"A":"-", "B":":"}
colors = {"1":1, "2":2}
cmap = {"colorbar_label":"jet"}
mne.viz.plot_compare_evokeds(evokeds, colors=colors, linestyles=linestyles, cmap=cmap)
```
(Mixed would depend on 1. giving a `cmap` parameter, 2. `colors` being a dict that links from HED tags to numeric.)

Mixed by order:

```
evokeds = [evoked1, evoked2, evoked3, evoked4]
linestyles = ["-", ":", "-", ":"]
colors = [1, 1, 2, 2]
cmap = {"colorbar_label":"jet"}
mne.viz.plot_compare_evokeds(evokeds, colors=colors, linestyles=linestyles, cmap=cmap)

```

(Currently, evokeds can be either 1. an Evoked obj, 2. a list of evoked objs, 3. a dict of evoked objs, 4. a list of lists of evoked objs, 5. a dict of lists of evoked objs.)

Right now I'm not sure the "dict of dicts" approach we discussed would be such a good idea.
`colors = {"condition_label":[{"1":1}, {"2":2}, {"3":3}]}` doesn't look so good.
Creating a pandas series might help, but isn't it overkill when we could simply have a dict?

I think your 'meta' idea would be better for something where we allow more flexibility (e.g., seaborn which has hue, col, row, marker, ...). But here we only ever have two: color and linestyle. Right?

agramfort commented:
@jona-sassenhagen fine with the proposal except I don't like the:

cmap = {"colorbar_label":"jet"}

cmap should be None, str or an instance of plt.cm.XXX

so ok for cmap = 'jet'

so colors and linestyles cannot be dict if evoked is not a dict. If evoked is a dict then everything is a dict and we use HED

last thing, how do you specify what to show in the legend?

agramfort commented:
sorry @jona-sassenhagen I was not notified of your edit of the previous message.

I don't understand why you say that the mixed use case will not work.

My summary is:

List input for Continuous

```
evokeds = [evoked1, evoked2, evoked3, evoked4]
colors = list(range(len(evokeds)))
cmap = {"colorbar_label":"viridis"}
mne.viz.plot_compare_evokeds(evokeds, colors=colors, cmap=cmap)
```
`colors` is of the same shape as `evokeds`

then we use evokeds as dict:

Factorial:

```
evokeds = {'A/C': evoked1, 'B/C': evoked2, 'A/C': evoked3, 'B/D': evoked4}
mne.viz.plot_compare_evokeds(evokeds, colors="viridis")
or
mne.viz.plot_compare_evokeds(evokeds, colors={"A":"r", "B":"b"}, linestyles={"C":"-", "D":":"})
```

and then mixed if one of the value in the color dict is not a string.

Mixed:
```
evokeds = {'A/1': evoked1, 'A/2': evoked2, 'B/1': evoked3, 'B/2': evoked4}
linestyles = {"A":"-", "B":":"}
colors = {"1":1, "2":2}
cmap = {"colorbar_label":"jet"}
mne.viz.plot_compare_evokeds(evokeds, colors=colors, linestyles=linestyles, cmap=cmap)
```

it means that what is not color coded with an int does not generate a colorbar.

clear? do I miss something?

jona-sassenhagen commented:
> I don't understand why you say that the mixed use case will not work.

I only meant that for
linestyles = ["-", ":", "-", ":"]
colors = [1, 1, 2, 2]
you don't know what the labels for the linestyle legend should be. So you'd have to omit the legend.

I'm fine with it as is otherwise, although this

evokeds = {'A/1': evoked1, 'A/2': evoked2, 'B/1': evoked3, 'B/2': evoked4}
colors = {"1":1, "2":2}
cmap = {"colorbar_label":"jet"}
mne.viz.plot_compare_evokeds(evokeds, colors=colors, cmap=cmap)
would mean you'd get a lot of

colors = {cond.split("/")[-1]:int(cond.split("/")[-1]) for cond in evokeds}
which would be a bit redundant.

agramfort commented:
>
> I only meant that for
> linestyles = ["-", ":", "-", ":"]
> colors = [1, 1, 2, 2]
> you don't know what the labels for the linestyle legend should be. So
> you'd have to omit the legend.
>

ok use all condition attributes as legend so you would have 4 entries in
legend.


> I'm fine with it as is otherwise, although this
>
> evokeds = {'A/1': evoked1, 'A/2': evoked2, 'B/1': evoked3, 'B/2': evoked4}
> colors = {"1":1, "2":2}
> cmap = {"colorbar_label":"jet"}
> mne.viz.plot_compare_evokeds(evokeds, colors=colors, cmap=cmap)
> would mean you'd get a lot of
>
> colors = {cond.split("/")[-1]:int(cond.split("/")[-1]) for cond in
> evokeds}
> which would be a bit redundant.


I don't see (yet?) the problem.

sounds like we have a plan. Unfortunately my 1,5 day of potential coding is
over
so we need volunteer or this will wait...

jona-sassenhagen commented:
> sounds like we have a plan. Unfortunately my 1,5 day of potential coding is
> over
> so we need volunteer or this will wait...

What do you mean, I'll do it :) I need this code.

jona-sassenhagen commented:
Ok I'm back at this since #4289 looks like it'll get merged soon.

agramfort commented:
tell me when it's not "Very WIP" anymore :)

jona-sassenhagen commented on mne/viz/tests/test_evoked.py:
```diff
@@ -117,56 +130,50 @@ def test_plot_evoked():
-        # test a bunch of params at once
+        for evokeds_ in (evoked.copy().pick_types(meg='mag'), contrast,
+                         [red, blue], [[red, evoked], [blue, evoked]]):
+            plot_compare_evokeds(evokeds_, picks=0)  # also tests CI
```
ha, much more efficient ...

jona-sassenhagen commented:
Closed for #4526
